crud/common/call.lua

```diff
@@ -53,101 +113,33 @@ local function call_impl(vshard_call, func_name, func_args, opts)
-                    else
-                        err = NotInitializedError:new("Function %s is not registered", func_name)
-                    end
-                end
-            end
-            err = errors.wrap(err)
-            return nil, CallError:new(utils.format_replicaset_error(
-                replicaset_uuid, "Function returned an error: %s", err
-            ))
+            return nil, wrap_vshard_err(err, func_name, replicaset_uuid)
         end
+
         results[replicaset_uuid] = result[1]
     end
 
     return results
 end
 
---- Calls specified function on all cluster storages.
---
--- Allowed functions to call can be specified by `crud.register` call.
--- If function with specified `opts.func_name` isn't registered,
--- global function with this name is called.
---
--- Uses vshard `replicaset:callrw`
---
--- @function rw
---
--- @param string func_name
---  A function name
---
--- @param ?table func_args
---  Array of arguments to be passed to the function
---
--- @tparam table opts Available options are:
---
--- @tparam ?number opts.timeout
---  Function call timeout
---
--- @tparam ?table opts.replicasets
---  vshard replicasets to call the function.
---  By default, function is called on the all storages.
---
--- Returns map {replicaset_uuid: result} with all specified replicasets results
---
--- @return[1] table
--- @treturn[2] nil
--- @treturn[2] table Error description
---
-function call.rw(func_name, func_args, opts)
-    return call_impl('callrw', func_name, func_args, opts)
-end
-
---- Calls specified function on all cluster storages.
---
--- The same as `rw`, but uses vshard `replicaset:callro`
---
--- @function ro
---
-function call.ro(func_name, func_args, opts)
-    return call_impl('callro', func_name, func_args, opts)
-end
+function call.single(bucket_id, func_name, func_args, opts)
+    dev_checks('number', 'string', '?table', {
+        mode = 'string',
+        prefer_replica = '?boolean',
+        balance = '?boolean',
+        timeout = '?number',
+    })
 
---- Calls specified function on a node according to bucket_id.
---
--- Exactly mimics the contract of vshard.router.callrw, but adds
--- better error hangling
---
--- @function rw_single
---
-function call.rw_single(bucket_id, func_name, func_args, options)
-    local res, err = vshard.router.callrw(bucket_id, func_name, func_args, options)
-
-    -- This is a workaround, until vshard supports telling us where the error happened
-    if err ~= nil then
-        if type(err) == 'table' and err.type == 'ClientError' and type(err.message) == 'string' then
-            if err.message == string.format("Procedure '%s' is not defined", func_name) then
-               err = NotInitializedError:new("crud isn't initialized on replicaset")
-            end
-        end
+    local vshard_call_name = get_vshard_call_name(opts.mode, opts.prefer_replica, opts.balance, opts.mode)
```

Comment:
So... I'm not sure that you should do it in such way.

Seems you can use just vshard.router.call('name', {mode=, balance=, prefer_replica=}, args, opts).
So you don't need to perform such logic.

(https://www.tarantool.io/en/doc/latest/reference/reference_rock/vshard/vshard_api/#lua-function.vshard.router.call)

Reply:
It wouldn't work for `replicaset:call`, so we decided to don't change it now.